server/manager/src/main/java/org/apache/accumulo/manager/BalanceManager.java

```diff
@@ -0,0 +1,331 @@
+        continue;
+      }
+
+      if (dl == Ample.DataLevel.USER
+          && !partitionedMigrations.get(Ample.DataLevel.METADATA).isEmpty()) {
+        log.debug("Not balancing {} because {} has migrations", dl, Ample.DataLevel.METADATA);
+        continue;
+      }
+
+      // Create a view of the tserver status such that it only contains the tables
+      // for this level in the tableMap.
+      SortedMap<TServerInstance,TabletServerStatus> tserverStatusForLevel =
+          createTServerStatusView(dl, manager.tserverStatus);
+      // Construct the Thrift variant of the map above for the BalancerParams
+      final SortedMap<TabletServerId,TServerStatus> tserverStatusForBalancerLevel = new TreeMap<>();
+      tserverStatusForLevel.forEach((tsi, status) -> tserverStatusForBalancerLevel
+          .put(new TabletServerIdImpl(tsi), TServerStatusImpl.fromThrift(status)));
+
+      log.debug("Balancing for tables at level {}", dl);
+
+      SortedMap<TabletServerId,TServerStatus> statusForBalancerLevel =
+          tserverStatusForBalancerLevel;
+      params =
+          BalanceParamsImpl.fromThrift(statusForBalancerLevel, manager.tServerGroupingForBalancer,
+              tserverStatusForLevel, partitionedMigrations.get(dl), dl, getTablesForLevel(dl));
+      wait = Math.max(tabletBalancer.balance(params), wait);
+      long migrationsOutForLevel = 0;
+      try (var tabletsMutator = getContext().getAmple().conditionallyMutateTablets(result -> {})) {
+        for (TabletMigration m : checkMigrationSanity(statusForBalancerLevel.keySet(),
+            params.migrationsOut(), dl)) {
+          final KeyExtent ke = KeyExtent.fromTabletId(m.getTablet());
+          if (partitionedMigrations.get(dl).contains(ke)) {
+            log.warn("balancer requested migration more than once, skipping {}", m);
+            continue;
+          }
+          migrationsOutForLevel++;
+          var migration = TabletServerIdImpl.toThrift(m.getNewTabletServer());
+          tabletsMutator.mutateTablet(ke).requireAbsentOperation()
+              .requireCurrentLocationNotEqualTo(migration).putMigration(migration)
+              .submit(tm -> false);
+          log.debug("migration {}", m);
+        }
+      }
+      totalMigrationsOut += migrationsOutForLevel;
+
+      // increment this at end of loop to signal complete run w/o any continue
+      levelsCompleted++;
+    }
+    final long totalMigrations =
+        totalMigrationsOut + partitionedMigrations.values().stream().mapToLong(Set::size).sum();
+    balancerMetrics.assignMigratingCount(() -> totalMigrations);
+
+    if (totalMigrationsOut == 0 && levelsCompleted == Ample.DataLevel.values().length) {
+      synchronized (balancedNotifier) {
+        balancedNotifier.notifyAll();
+      }
+    } else if (totalMigrationsOut > 0) {
+      manager.nextEvent.event("Migrating %d more tablets, %d total", totalMigrationsOut,
```

Comment:
All of the code in the class was copied as is from the manager and then some small changes were made to add the `manager.` prefix here.  Did not change the functionality of any of the code though.